src/app/application/services/data-source/Router.ts

```diff
@@ -0,0 +1,21 @@
+export default class Router<T> {
+  routes: Map<URLPattern, T> = new Map()
+  constructor(routes: Record<string, T>) {
+    Object.entries(routes).forEach(([key, value]) => {
+      this.routes.set(new URLPattern({
+        pathname: key,
+      }), value)
+    })
+  }
+
+  match(path: string) {
+    for (const [pattern, route] of this.routes) {
+      const _url = `data:${path}`
+      if (pattern.test(_url)) {
+        const args = pattern.exec(_url)
+        return { route, params: args?.pathname.groups }
+      }
+    }
+    throw new Error(`Matching route for '${path}' not found`)
+  }
+}
```

Comment:
We use this class currently in testing, and I might have edited it slightly since I wrote this. They'll both need matching up and reusing instead of having two separate Router classes.

Reply:
Maybe a unit test of this would be useful. This code is a little hairy and ensuring it works as expected will make things easier to debug